contributors/design-proposals/deployment.md

```diff
@@ -159,56 +165,89 @@ https://github.com/kubernetes/kubernetes/issues/11234 tracks updating kubelet
+of existing ReplicaSets. This is similar to how it is already used by DaemonSets.
+
+TemplateGeneration is not authoritative and only helps in constructing the name for the new
+ReplicaSet in case there is no other ReplicaSet that matches the Deployment. The Deployment
+controller still decides what is the new ReplicaSet by comparing PodTemplateSpecs.
+If no matching ReplicaSet is found, the controller will try to create a new ReplicaSet using
+its current TemplateGeneration.
+
+The naming scheme used by ReplicaSets (*deployment.name-podtemplatehash*) will need to change
+to something different because we cannot migrate old ReplicaSets to use TemplateGeneration
+for their names and we want to avoid name collisions. For new RS names, we can either append
+the TemplateGeneration to the Deployment name or we can compute the hash of
+*deployment.name+deployment.templategeneration*, extend its size deterministically in different
+ways to ensure that a new RS name will never collide with an old one, and use that as the hash
+that shall be appended in the Deployment name to produce the new ReplicaSet name.
+
+
 ## Changing Deployment mid-way
 
 ### Updating
 
-Users can update an ongoing deployment before it is completed.
-In this case, the existing deployment will be stalled and the new one will
+Users can update an ongoing Deployment before it is completed.
+In this case, the existing rollout will be stalled and the new one will
 begin.
-For ex: consider the following case:
-- User creates a deployment to rolling-update 10 pods with image:v1 to
+For example, consider the following case:
+- User updates a Deployment to rolling-update 10 pods with image:v1 to
   pods with image:v2.
-- User then updates this deployment to create pods with image:v3,
-  when the image:v2 RC had been ramped up to 5 pods and the image:v1 RC
+- User then updates this Deployment to create pods with image:v3,
+  when the image:v2 RS had been ramped up to 5 pods and the image:v1 RS
   had been ramped down to 5 pods.
-- When Deployment Controller observes the new deployment, it will create
-  a new RC for creating pods with image:v3. It will then start ramping up this
-  new RC to 10 pods and will ramp down both the existing RCs to 0.
+- When Deployment Controller observes the new update, it will create
+  a new RS for creating pods with image:v3. It will then start ramping up this
+  new RS to 10 pods and will ramp down both the existing RSs to 0.
 
 ### Deleting
 
-Users can pause/cancel a deployment by deleting it before it is completed.
-Recreating the same deployment will resume it.
-For ex: consider the following case:
-- User creates a deployment to rolling-update 10 pods with image:v1 to
+Users can cancel a rollout by doing a non-cascading deletion of the Deployment
+it before it is complete. Recreating the same Deployment will resume it.
+For example, consider the following case:
+- User creates a Deployment to rolling-update 10 pods with image:v1 to
```

Comment:
User creats a Deployment to perform a rolling-update for 10 pods from image:v1 to image:v2.